src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -101,9 +104,19 @@ def resolve(
             raise error from e
 
         req_set = RequirementSet(check_supported_wheels=check_supported_wheels)
-        for candidate in result.mapping.values():
+        # sort to ensure base candidates come before candidates with extras
+        for candidate in sorted(result.mapping.values(), key=lambda c: c.name):
```

Comment:
Hmm, I’m not sure if sorting the entire list would break some things (we are generally quite delibrate to keep the input order in most cases). But I guess if it’s otherwise too cumbersome to sort a base candidate before extras we can just do this for now until (if) anyone complains.

Reply:
You make a good point. I don't think the order is very important at this late stage in the process, but best not to mess with it too much just in case. I changed this to a more conservative sort key in 21bfe40. I think this could be a nice middle ground: it only moves the extras to the rear, which should hardly affect anything. What do you think?

Reply:
Sounds like a good middle ground if it does not affect performance too much. (Another approach I was considering is to loop through the list once to process the non-extras ones, while collecting extras to be processed in a second loop.)

Reply:
I was going to argue with respect to readability but now that I look it over once more I actually prefer the two-pass approach you suggest. Since we already branch on the presence of the extra, there's really no reason to keep it in the same loop. So I'll make the change.

Out of curiosity, I did run a quick test on the performance difference. I would have expected it to be negligible but this snippet seems to indicate that there is a slight performance hit for the sorting approach, though it's difficult to make it completely representative. It's still in the same order of magnitude though, which is all that matters for many practical cases.

```python
import random
import timeit


nb_elements: int = 10_000
nb_iterations: int = 10_000

l = random.choices([0, 1], weights=[9, 1], k=nb_elements)


def with_sort():
    for i in sorted(l, key=lambda i: i):
        pass


def pass_twice():
    for i in l:
        if i == 0:
            pass
    for i in l:
        if i == 1:
            pass


def pass_twice_cache():
    second_pass: list[int] = []
    for i in l:
        if i == 0:
            pass
        else:
            second_pass.append(i)
    for i in second_pass:
        pass


print("with_sort", timeit.timeit(with_sort, number=nb_iterations))
print("pass_twice", timeit.timeit(pass_twice, number=nb_iterations))
print("pass_twice_cache", timeit.timeit(pass_twice_cache, number=nb_iterations))
```

result:
```
with_sort 7.437978116000522
pass_twice 4.571586305999517
pass_twice_cache 2.8610327940004936  # difference becomes less pronounced as the weights near 50-50
```

Reply:
> Since we already branch on the presence of the extra

This is not right, sorry about that. We branch on the presence of the extra iff there is no ireq. Extras candidates with an ireq require the same logic as base candidates with an ireq, so unless you think the performance difference is sufficiently significant to warrant the change I'm inclined to leave it as is after all.

This also means that the "middle ground" still reorders a bit more than I initially thought (because not all extras candidates have a corresponding base, meaning their processing may shift to the rear). Let me know if you think this is unacceptable, then I'll make an effort to implement smarter (but more complex) sorting.

Reply:
I don’t think performance is the issue here, the thing I’m worried about is mostly doing to much reordering may affect semantics, as mentioned above. I’m willing to opt for the easiest logic possible and see if it really affects anyone in practice.